Add application for load testing

Nico-Sanchez opened the pull request:
Add new ArenaLoadTest application under umbrella project to stress-test the Arena backend architecture.

- Add new application under our umbrella project.
- Add deploy scripts for the ArenaLoadTest app.
- Add respective README markdown.
- Start this new application when starting the umbrella project.
- Add dummy behavior for the mocked clients that join the different games.
  - Will use bots from the bots application in the future (not developed yet).
- Change credo TestFileExtension check due to app named *_test.ex

Closes #371
Next iterations will do the rest of #361 features.

lotuuu commented on apps/arena/lib/arena/game/player.ex:
```diff
@@ -279,9 +279,9 @@ defmodule Arena.Game.Player do
-  defp skill_key_execution_action(skill_key) do
-    "EXECUTING_SKILL_#{String.upcase(skill_key)}" |> String.to_existing_atom()
-  end
+  defp skill_key_execution_action("1"), do: :EXECUTING_SKILL_1
```
These changes are already included in #367 🤔

Nico-Sanchez replied:
Yeah but I need it for the mocked players' actions. Otherwise the app crashes when game starts.

lotuuu replied:
Shouldn't this PR be pointed towards that one then? Either that, or we close #367 when we merge this

Nico-Sanchez replied:
Let's see which gets merged first 👀

agustinesco replied:
#367 won

lotuuu commented on apps/arena_load_test/lib/arena_load_test.ex:
```diff
@@ -0,0 +1,5 @@
+defmodule ArenaLoadTest do
```
We have an empty boilerplate module here

Nico-Sanchez replied:
Yeah. Other apps have it as well that's why I didn't delete it.
I can delete it if you think is better to do so.

lotuuu replied:
Yes! We can bring it back if we need it later

Nico-Sanchez replied:
Done!

lotuuu commented on Makefile:
```diff
@@ -51,3 +51,10 @@ generate-game-client-protos:
 		--proto_path=apps/serialization \
 		messages.proto
+
+generate-arena-load-test-protos:
```
Shouldn't we add this new command to `generate-protos`?

Nico-Sanchez replied:
Yeah. done!

Nico-Sanchez commented:
> I'm getting a weird error when i try to run tests with more than 99 client (i didn't tested the exact case where this starts to happen)
> 
> ```
> [info] Client waiting for game to join
> ** (IEx.Error) error when printing EXIT message: :undef
> 
> Interactive Elixir (1.16.1) - press Ctrl+C to exit (type h() ENTER for help)
> *** ERROR: Shell process terminated! (^G to start new job) ***
> DEFAULT FORMATTER CRASHED
> DEFAULT FORMATTER CRASHED
> DEFAULT FORMATTER CRASHED
> DEFAULT FORMATTER CRASHED
> DEFAULT FORMATTER CRASHED
> ```
> 
> Also when the tests are running we can see a lot of debug messages, is this intended? if we want to check that they are running maybe we should parse this message to something readable
> 
> ![image](https://private-user-images.githubusercontent.com/106101218/312562452-892efed2-2459-4431-8e07-ce90dd7d234d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzNTE2NzQsIm5iZiI6MTcxMDM1MTM3NCwicGF0aCI6Ii8xMDYxMDEyMTgvMzEyNTYyNDUyLTg5MmVmZWQyLTI0NTktNDQzMS04ZTA3LWNlOTBkZDdkMjM0ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMxM1QxNzM2MTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZDcyZTZmOTllNGY5ZGQ1NzNmNjQ3MWRlYzk2NmNhYzZlMWQ5ZmY4MmRkMTA2ODkzMDNjNDY4N2E3MTcwYzllJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.SgoGMgG5h0mmwUWhb7GUlzY6PJawX8A1uLorBoCaPg4)

Nice catch!
About the crash with more than 100 clients, you have to increase the number of file descriptors a process can consume. When you try to open more websockets than FD's your SO allows, it crashes.
You can increase the FD's by running (e.g. 10000 file descriptors):
`ulimit -n 10000`
I added this to the README file.

And about the second problem, I deleted the debug: :trace flag to avoid flooding the logs.

agustinesco reviewed (dismissed):
LGTM! tested the load tests multiple times with distinct amount of players and it worked like a charm

lotuuu commented on apps/arena_load_test/lib/arena_load_test/game_socket_handler.ex:
```diff
@@ -0,0 +1,109 @@
+
+  # Callbacks
+  def handle_frame({_type, _msg} = _frame, state) do
+    # Logger.info("Received frame with msg: #{_msg}")
```
Let's remove this comment if we don't want it!

lotuuu commented on apps/arena_load_test/lib/arena_load_test/game_socket_handler.ex:
```diff
@@ -0,0 +1,109 @@
+  end
+
+  def handle_cast({:send, {_type, _msg} = frame}, state) do
+    # Logger.info("Sending frame with payload: #{msg}")
```
Same!

Nico-Sanchez replied:
Done!

lotuuu commented on apps/arena_load_test/lib/arena_load_test/socket_handler.ex:
```diff
@@ -0,0 +1,71 @@
+  # Private
+  defp ws_url(player_id) do
+    host = SocketSupervisor.server_host()
+    # TODO must replace character with a random available one.
```
Why isn't this done?

Nico-Sanchez replied:
Because its part of #361 
Added its mention so its more clear.

lotuuu reviewed (changes requested):
Feature works great, marked some code design thingies!

Nico-Sanchez commented on apps/arena_load_test/lib/arena_load_test/game_socket_handler.ex:
```diff
@@ -0,0 +1,109 @@
+  defp ws_url(client_id, game_id) do
+    host = SocketSupervisor.server_host()
+
+    case System.get_env("SSL_ENABLED") do
```
I don't think it adds clarity but I'll keep in mind for future PRs.

Nico-Sanchez commented on apps/arena_load_test/lib/arena_load_test/socket_handler.ex:
```diff
@@ -0,0 +1,71 @@
+    # Enum.random(["h4ck", "muflus"])
+    character = "h4ck"
+
+    case System.get_env("SSL_ENABLED") do
```
Answered!

agustinesco reviewed (approved):
LGTM! 🎸
